pkg/server/server_controller_orchestration.go

```diff
@@ -204,39 +212,64 @@ func (c *serverController) startControlledServer(
 	}
 
 	topCtx := ctx
+
 	// Use a different context for the tasks below, because the tenant
 	// stopper will have its own tracer which is incompatible with the
 	// tracer attached to the incoming context.
-
 	tenantCtx := logtags.AddTag(context.Background(), "tenant-orchestration", nil)
+	tenantCtx = logtags.AddTag(tenantCtx, "tenant", tenantName)
 
 	tenantStopper := stop.NewStopper()
 
+	// shutdownInterface enables the propagate-close task to
+	// initiate a graceful drain on the server started by the other
+	// task below.
+	var shutdownInterface tenantServerShutdownInterface
+
 	// Ensure that if the surrounding server requests shutdown, we
 	// propagate it to the new server.
 	if err := c.stopper.RunAsyncTask(ctx, "propagate-close", func(ctx context.Context) {
 		select {
 		case <-tenantStopper.ShouldQuiesce():
 			// Tenant is terminating on their own; nothing else to do here.
 			log.Infof(ctx, "tenant %q terminating", tenantName)
+
 		case <-c.stopper.ShouldQuiesce():
 			// Surrounding server is stopping; propagate the stop to the
 			// control goroutine below.
+			// Note: we can't do a graceful drain in that case because
+			// the RPC service in the surrounding server may already
+			// be unavailable.
 			log.Infof(ctx, "server terminating; telling tenant %q to terminate", tenantName)
 			tenantStopper.Stop(tenantCtx)
+
 		case <-stopRequestCh:
-			// Someone requested a shutdown.
+			// Someone requested a graceful shutdown.
 			log.Infof(ctx, "received request for tenant %q to terminate", tenantName)
+
+			// Ensure that the graceful drain for the tenant serer aborts
+			// early if the Stopper for the surrounding server is
+			// prematurely shutting down. This is because once the surrounding node
+			// starts quiescing tasks, it won't be able to process KV requests
+			// by the tenant server any more.
+			drainCtx, cancel := c.stopper.WithCancelOnQuiesce(tenantCtx)
+			defer cancel()
+			shutdownInterface.maybeCallDrain(drainCtx)
```

Comment:
I think we probably have other problems of this nature so perhaps it isn't worth fixing, but I think the current structure will fail to drain the server if the following sequence happens:

1) The propogate-close task starts and is running
2) The managed-tenant-server task starts and reaches line 349 but has not run line 350
3) createServerEntryLocked returns the server entry, the entry is added to the server slice, the severs slice mutex is unlocked.
4) requestAllStopped() is called.
5) propogate-close sees the stoprequest, and runs myabeCallDrain(), but we haven't set the server yet, so it reports nothing to drain, even though our server already has drainable components started.

I think this is minor because the server won't be serving any user connections yet (since the orchestrator doesn't think it is started). So it just means any jobs or other internal components might not get drained.